repair/repair.cc

```diff
@@ -1920,20 +1922,57 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            std::unordered_set<gms::inet_address> dc_nodes;
+            sstring ks_source_dc;
+            if (ignored_in_my_dc < 0) {
+                size_t ignored_in_my_dc = 0;
+                topology.for_each_node([&] (const locator::node* node) {
+                    if (node->dc_rack().dc != source_dc) {
+                        return;
+                    }
+                    if (ignore_nodes.contains(node->endpoint())) {
+                        ++ignored_in_my_dc;
+                    } else {
+                        dc_nodes.insert(node->endpoint());
+                    }
+                });
+            }
+            // Allow repairing in the dc only if there are enough replicas
+            if (!source_dc.empty() && !dc_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    if (ignored_in_my_dc) {
+                        // Don't take any chances if there are ignored nodes in this dc
+                        break;
+                    }
+                    auto opts = strat.get_config_options();
+                    auto rf = opts.contains(source_dc) ? std::atoi(opts[source_dc].c_str()) : 0;
+                    if (rf > 1) {
+                        ks_source_dc = source_dc;
```

Comment:
ks_source_dc is never different than source_dc, so it would be enough to just clear source_dc if we decide to use all dcs and not introduce new state.

Reply:
ks_source_dc may be either empty or equal to source_dc.
The reason it's per keyspace is that different keyspace have different replication strategies and configuration so it might be possible to use source_dc for some keyspaces but not for others.